DNM Design For OADP CLI NVP (#1797)
openshift/oadp-operator: docs/design/kubectl-oadp.md

Background
The current OADP cli is suboptimal as oc backup delete $foo deletes the k8 object instead of the backup but velero backup delete $foo deletes the backup, along with the backup files in storage. Currently, customers would need to alias velero in order to delete their backups, which is not ideal. The purpose of kubectl-oadp would be to make the cli experience better and easier to use along with enabling users to be able to get the logs of the backups. 

Goals
- Customers can create backups and restores
- A non-cluster admin can create Non-Admin-Backups (NAB)

High-Level Design
Creating a kubectl plugin (kubctl-oadp) will be a good solution to the problem at hand. It will be able to create/delete backups and restores. Non-cluster admin will be able to create NABs without the need for cluster admin to do it for them. 

Detailed Design
The kubectl plugin will have imports from velero to help with the creation/deletion of backups and restores. It will be written in Golang and be using cobra for command-line parsing. The non-admin cli can be a subset of some backup clis that already exist such as backup.go and create.go. 

CLI Examples

> oc oadp backup create 
> oc oadp backup logs
> oc oadp restore create
> oc oadp restore logs 

Reviewer: Would be nice to have some markdown `code`  decorations for the code here.

````suggestion
```sh
oc oadp backup create 
oc oadp backup logs
oc oadp restore create
oc oadp restore logs 
```
````

Author: fixed

# Kubectl-oadp plugin design

## Abstract

> The purpose of this plugin is to allow the customer to create and delete backups, along with creating restores in OADP without needing to alias velero to do so.

Author: Updated to be more clear

## Background
The current OADP cli is suboptimal as oc backup delete $foo deletes the k8 object instead of the backup but velero backup delete $foo deletes the backup, along with the backup files in storage. Currently, customers would need to alias velero in order to delete their backups, which is not ideal. The purpose of kubectl-oadp would be to make the cli experience better and easier to use along with enabling users to be able to get the logs of the backups.

## Goals
- Customers can create and delete backups
- A non-cluster admin can create Non-Admin-Backups (NAB)

## Non-Goals

> - Non-Admin-Restore due to time constraints

Reviewer 2: NABSL and other Non-Admin CRs are also out of scope, right ?

Author: Correct, fixed code to reflect this

> 

Reviewer 2: Lets add a use-case section too

Author: fixed

## High-Level Design

> Creating a kubectl plugin (kubectl-oadp) will be a good solution to the problem at hand. It will be able to create/delete backups and restores. Non-cluster admin will be able to create NABs without the need for cluster admin to do it for them. 

Reviewer 3: @mpryc From what @weshayutin said above, I think we're *only* acting on NABs and not Velero backups. Creating the NAB/NAR/etc. and the Non-admin controller creates the velero CRs.

> ## Detailed Design

Author: I will leave it out of the design doc for now, but using krew for installations would be nice in the future

The kubectl plugin will have imports from velero to help with the creation/deletion of backups and restores. It will be written in Golang and be using cobra for command-line parsing. The non-admin cli can be a subset of some backup clis that already exist such as backup.go and create.go. 

What we discovered with the regular commands such as version, backup, and restore we can just import the libraries 

> ```sh

Reviewer 2: s/sh/go

Author: fixed, changed sh to go

package cmd

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"
	"github.com/vmware-tanzu/velero/pkg/cmd/cli/backup"
	"github.com/vmware-tanzu/velero/pkg/cmd/cli/restore"
	"github.com/vmware-tanzu/velero/pkg/cmd/cli/version"
)
```
With non-admin, we would have to create the cli ourselves since there are no cli’s for it.

```sh
func NewCreateCommand(f client.Factory, use string) *cobra.Command {
	o := NewCreateOptions()

	c := &cobra.Command{
		Use:   use + " NAME",
		Short: "Create a non-admin backup",
		Args:  cobra.MaximumNArgs(1),
		Run: func(c *cobra.Command, args []string) {
			cmd.CheckError(o.Complete(args, f))
			cmd.CheckError(o.Validate(c, args, f))
			cmd.CheckError(o.Run(c, f))
		}, 

```
CLI Examples
```sh
oc oadp backup create 
oc oadp backup logs
oc oadp restore create

> oc oadp restore logs 

Reviewer: heh

> oc oadp restore logs 

Author: fixed. added more cli

```

## Alternatives Considered
An alternative that was considered was creating our own CLI from scratch and not using a plugin. We can instead use the existing oc commands and just add on to them with a kubectl plugin. 

Aliasing is another way in which you could access the velero command line. However, this is not ideal because some individuals do not have permission to use the velero cli, so the kubectl plugin would allow those people to use velero cli. 

## Security Considerations

> The user first enters in a command in which the plugin reads the command and creates a velero client factory. The client factory then creates a connection to the kubernetes server.

Author: Revamped security considerations, now more concise to reader

## Compatibility

> This plugin would need to be updated so that it would be importing the right version of the velero backup and restore libraries. 

Author: Added Future work paragraph

# Kubectl-oadp plugin design

## Abstract

> The purpose of this plugin is to allow the customer to delete backups, along with creating restores in OADP without needing to alias velero to do so.

Reviewer 3: ...to allow the customer to create and delete backups...

## Background
The current OADP cli is suboptimal as oc backup delete $foo deletes the k8 object instead of the backup but velero backup delete $foo deletes the backup, along with the backup files in storage. Currently, customers would need to alias velero in order to delete their backups, which is not ideal. The purpose of kubectl-oadp would be to make the cli experience better and easier to use along with enabling users to be able to get the logs of the backups.

## Goals
- Customers can create backups and restores

> - A non-cluster admin can create Non-Admin-Backups (NAB)

Reviewer 3: Would the non-admin backup/restore API calls be different? Since the CRDs are different, I imagine those would need different APIs here as well.

## Non-Goals
- Non-Admin-Restore due to time constraints

## High-Level Design

> Creating a kubectl plugin (kubctl-oadp) will be a good solution to the problem at hand. It will be able to create/delete backups and restores. Non-cluster admin will be able to create NABs without the need for cluster admin to do it for them. 

Author: fixed

# Kubectl-oadp plugin design

## Abstract

> The purpose of the kubectl-oadp plugin is to allow the customer to create and delete backups, along with creating restores in OADP without needing to alias velero to do so. Non-cluster admins should also be able to create NABs and get the logs from them.

Reviewer 2: NIT: NABs (Non-Admin Backups)

Author: fixed thanks!

## Background

> The current OpenShift cli is suboptimal as oc backup delete $foo deletes the k8 object instead of the backup but velero backup delete $foo deletes the backup, along with the backup files in storage. Currently, customers would need to alias velero in order to delete their backups, which is not ideal. The purpose of kubectl-oadp would be to make the cli experience better and easier to use along with enabling users to be able to get the logs of the backups.

Reviewer: ```suggestion
The current OpenShift cli is suboptimal as oc backup delete $foo deletes the k8s object instead of the backup but velero backup delete $foo deletes the backup, along with the backup files in storage. Currently, customers would need to alias velero in order to delete their backups, which is not ideal. The purpose of kubectl-oadp would be to make the cli experience better and easier to use along with enabling users to be able to get the logs of the backups.
```

## Goals

> - Customers can create, delete, and restore backups

Reviewer 2: backup logs too, right ?

Author: fixed

> - Customers can create, delete, and restore backups

Reviewer 4: The goals section should talk about backup and restores as separate resources perhaps?
Actions associated with each CR

Author: Listed as separate resources now, thanks

> - A non-cluster admin can create, delete and recieve the logs of the Non-Admin-Backups (NAB)

Author: fixed

## Non-Goals
- Non-Admin-Restore and other Non-Admin CRs due to time constraints

## Use-Case

> A use case of the kubectl-oadp plugin could be when a non-cluster admin would like to create a NAB or view the logs of a NAB without having to depend on the cluster admins to do so. Another use case would be if a developer would want to create a normal backup, they can just use this plugin to do so.

Reviewer 2: NIT: s/plugin/cli

Author: fixed

> ## High-Level Design

Author: added in high-level design

> Creating a kubectl plugin (kubectl-oadp) will be a good solution to the problem at hand. It will be able to create/delete backups and restores. Non-cluster admin will be able to create NABs without the need for cluster admin to do it for them. A way to distinguish between creating either NABs or regular backups would be in the cli. For intstance, if you would like to create a NAB, you would have to do kubectl oadp create nonadmin backup [backupname].  

Reviewer 2: oadp create nonadmin backup [backupname] <-> oadp nonadmin backup create [backupname]

Author: fixed

> Creating a kubectl plugin (kubectl-oadp) will be a good solution to the problem at hand. It will be able to create/delete backups and restores. Non-cluster admin will be able to create NABs without the need for cluster admin to do it for them. A way to distinguish between creating either NABs or regular backups would be in the cli. For intstance, if you would like to create a NAB, you would have to do kubectl oadp create nonadmin backup [backupname].  

Author: fixed

## Detailed Design

> The kubectl plugin will have imports from velero to help with the creation/deletion of backups and restores. It will be written in Golang and be using cobra for command-line parsing. The non-admin cli can be a subset of some backup clis that already exist such as backup.go and create.go. The plugin can be manually installed by utilizing the quick-create.sh file which creates the plugin and moves it to the respective location.

Reviewer 2: We can remove the install details of`quick-create.sh` from the design.

Reviewer 2: Also, I think its work exploring https://krew.sigs.k8s.io/ for plugin installation. 

We could just do: `kubectl krew install oadp`

Then just use the CLI: `kubectl oadp backup create`

- Plugin index: https://github.com/kubernetes-sigs/krew-index
- Dev guide: https://krew.sigs.k8s.io/docs/developer-guide/

This is similar to how Mac OS brew package manager works. Would be pretty neat and clean.

Author: Deleted quick-create.sh, it is worth exploring krew, but for now will keep it off design doc

What we discovered with the regular commands such as version, backup, and restore we can just import the libraries 

```go
package cmd

import (
	"fmt"
	"os"

	"github.com/spf13/cobra"
	"github.com/vmware-tanzu/velero/pkg/cmd/cli/backup"
	"github.com/vmware-tanzu/velero/pkg/cmd/cli/restore"
	"github.com/vmware-tanzu/velero/pkg/cmd/cli/version"
)
```

> With non-admin, we would have to create the cli ourselves since there are no cli’s for it.

Author: fixed

```go
func NewCreateCommand(f client.Factory, use string) *cobra.Command {
	o := NewCreateOptions()

	c := &cobra.Command{
		Use:   use + " NAME",
		Short: "Create a non-admin backup",
		Args:  cobra.MaximumNArgs(1),
		Run: func(c *cobra.Command, args []string) {
			cmd.CheckError(o.Complete(args, f))
			cmd.CheckError(o.Validate(c, args, f))
			cmd.CheckError(o.Run(c, f))
		}, 

```
CLI Examples
```sh

> kubectl oadp backup create

Reviewer 2: NIT: `kubectl oadp backup create <BACKUP_NAME> ...`

kubectl oadp backup delete 
kubectl oadp nonadmin backup logs
kubectl oadp nonadmin backup describe 
kubectl oadp restore create
```

## Alternatives Considered
An alternative that was considered was creating our own CLI from scratch and not using a plugin. We can instead use the existing oc commands and just add on to them with a kubectl plugin. 

> Aliasing is another way in which you could access the velero command line. However, this is not ideal because some individuals do not have permission to use the velero cli, so the kubectl plugin would allow those people to use velero cli. 

Reviewer: I don't understand this point. Could you clarify?
> some individuals do not have permission to use the velero cli

Reviewer: ```suggestion
Aliasing is another way in which you could access the velero command line. However, this is not ideal because non admins do not have permission to use the velero cli, so the kubectl plugin would allow non admins to retrieve their non admin backups similar to admins via velero cli. 
```

Author: changed code to clarify

## Security Considerations

> The security for the plugin is controlled by OpenShift RBAC, which allows cluster admins to control permissions of users below it. This is untilized to allows users to only be allowed to use the commands that they have permission to use by their namespace. The plugin would also create error statements like "Unautherized Access" whenever a user who does not have the right permissions tries to do unautherized commands.

Author: fixed
